In [1]:
%autosave 60

import tensorflow as tf
import numpy as np
from tensorflow.contrib.layers import batch_norm, dropout

Autosaving every 60 seconds


Exercise 8
===

Create a DQN to solve the "BipedalWalker-v2" gym.

In [2]:
import gym

env = gym.make("BipedalWalker-v2")
obs = env.reset()
print(obs)

[ 2.74738157e-03 -1.68065156e-05  1.30737595e-03 -1.59998858e-02
  9.19198319e-02 -1.72529754e-03  8.60291928e-01  2.67775295e-03
  1.00000000e+00  3.23249958e-02 -1.72517262e-03  8.53842065e-01
  1.22014224e-03  1.00000000e+00  4.40814197e-01  4.45820302e-01
  4.61422980e-01  4.89550382e-01  5.34103036e-01  6.02461278e-01
  7.09149182e-01  8.85932207e-01  1.00000000e+00  1.00000000e+00]


In [3]:
from matplotlib import pyplot as plt

%matplotlib inline

def render_env(env):
    plt.imshow(env.render(mode="rgb_array"))

#render_env(env)

In [4]:
print(env.action_space)
print(env.action_space.shape)
print(env.action_space.low)
print(env.action_space.high)

Box(4,)
(4,)
[-1 -1 -1 -1]
[1 1 1 1]


In [5]:
from tensorflow.contrib.layers import fully_connected

neurons_per_layer = [20]#[40, 40, 20]
learning_rate = 1 #0.001
n_discretized_actions = 3 # -1, -0.5, 0, +0.5, +1
n_actions = env.action_space.shape[0]
#n_outputs = env.action_space.shape[0]
n_outputs = n_discretized_actions ** n_actions
print("n_discretized_actions", n_discretized_actions, "n_actions", n_actions, "n_outputs", n_outputs)

def q_network(X_state, scope):
    current_layer = X_state
    layers = []
    with tf.variable_scope(scope) as scope:
        for neurons in neurons_per_layer:
            current_layer = fully_connected(current_layer, neurons)#, activation_fn=tf.nn.sigmoid)
            #print("current_layer shape", current_layer.get_shape())
        #print("input shape", current_layer.get_shape())
        outputs = fully_connected(current_layer, n_outputs, activation_fn=None)
    trainable_vars = tf.get_collection(tf.GraphKeys.TRAINABLE_VARIABLES, scope=scope.name)
    trainable_vars_by_name = {var.name[len(scope.name)]: var for var in trainable_vars}
    return outputs, trainable_vars_by_name

n_discretized_actions 3 n_actions 4 n_outputs 81


In [6]:
print(obs.shape)

(24,)


In [7]:
X_state = tf.placeholder(tf.float32, shape=[None, obs.shape[0]])
actor_q_values, actor_vars = q_network(X_state, "q_networks/actor")
#actor_greedy_action_index_summary = tf.summary.scalar("actor greedy action index", tf.argmax(actor_q_values))
actor_greedy_q_value_summary = tf.summary.scalar("actor greedy q value", tf.reduce_max(actor_q_values))
actor_summary = tf.summary.merge_all()
critic_q_values, critic_vars = q_network(X_state, "q_networks/critic")

copy_ops = [actor_var.assign(critic_vars[var_name])
            for var_name, actor_var in actor_vars.items()]
copy_critic_to_actor = tf.group(*copy_ops)

INFO:tensorflow:Summary name actor greedy q value is illegal; using actor_greedy_q_value instead.


In [8]:
#X_action = tf.placeholder(tf.int32, shape=[None, env.action_space.shape[0]])
X_action_index = tf.placeholder(tf.int32, shape=[None])
# This contains the Q-value for the actor's chosen action.
# q_value = tf.reduce_sum(critic_q_values * tf.one_hot(X_action, n_outputs), axis=1, keep_dims=True)
#q_value = tf.reduce_sum(critic_q_values * tf.one_hot(n_outputs, 1), axis=1, keep_dims=True)
q_value = tf.reduce_sum(critic_q_values * tf.one_hot(X_action_index, 1), axis=1, keep_dims=True)
#q_value = critic_q_values
print("critic_q_values.get_shape()", critic_q_values.shape, "q_value.get_shape()", q_value.get_shape())

critic_q_values.get_shape() (?, 81) q_value.get_shape() (?, 1)


In [9]:
# The Actor's reward + future discounted estimated Q-value.
y = tf.placeholder(tf.float32, shape=[None, 1])
cost = tf.reduce_mean(tf.square(y - q_value))
global_step = tf.Variable(0, trainable=False, name="global_step")
optimizer = tf.train.AdamOptimizer(learning_rate=learning_rate)
training_op = optimizer.minimize(cost, global_step=global_step)

init = tf.global_variables_initializer()
saver = tf.train.Saver()

In [10]:
from collections import deque

replay_memory_size = 10000
replay_memory = deque([], maxlen=replay_memory_size)

def sample_memory(batch_size):
    indices = np.random.permutation(len(replay_memory))[:batch_size]
    cols = [[], [], [], [], []] # state, action_index, reward, next_state, continue
    for idx in indices:
        memory = replay_memory[idx]
        for col, value in zip(cols, memory):
            col.append(value)
    cols = [np.array(col) for col in cols]
    return (cols[0], cols[1], cols[2].reshape(-1, 1), cols[3], cols[4].reshape(-1, 1))


In [11]:
from itertools import product
discretized_actions = [-1, 0, 1]
action_space_universe = np.array(list(product(discretized_actions, discretized_actions, discretized_actions, discretized_actions)))

print(action_space_universe.shape)
#print(action_space_universe)

def action_index_to_action(action_index):
    return action_space_universe[action_index]

(81, 4)


In [ ]:
# epsilon greed algorithm to explore the state-action space of the environment at the beginning.
def epsilon_greedy(q_values, step):
    eps_min = 0.05 # At the start 5% of the time the actor will choose the greedy action.
    eps_max = 1.0 # Once it has finished exploring, the actor will choose the greedy action 100% of the time.
    eps_decay_steps = 200000
    
    epsilon = max(eps_min, eps_max - (eps_max - eps_min) * step / eps_decay_steps)
    if np.random.rand() < epsilon:
        #return np.random.uniform(env.action_space.low, env.action_space.high, size=n_actions)
        return np.random.randint(len(action_space_universe))
    else:
        #return q_values.reshape(n_outputs)
        return np.argmax(q_values) # the optimal/greedy action.

In [ ]:
import os

def preprocess_observation(obs):
    # I don't think I need this method.
    return obs

n_steps = 210000
# Start training the critic DQN after 1000 game iterations.
# This has to be a lot bigger than the batch_size defined below.
training_start = 1000
training_interval = 3 # Run a training step every 3 game iterations start training_start.
save_steps = 50
copy_steps = 25 # Copy the critic to the actor every 25 training steps.
discount_rate = 0.99999 # 0.95
batch_size = 50
iteration = 0
checkpoint_path = "./BipedalWalker-v2.ckpt"
done = True # Environment needs to be reset
total_reward = 0
current_steps = 0

with tf.Session() as sess:
    file_writer = tf.summary.FileWriter("tf_logs", sess.graph)
    if os.path.isfile(checkpoint_path):
        print("Restoring checkpoint")
        saver.restore(sess, checkpoint_path)
    else:
        init.run()
    while True:
        step = global_step.eval()
        if step >= n_steps:
            break
        elif step != 0and step % 10000 == 0:
            print("global_step", step)
        iteration += 1
        if done:
            obs = env.reset()
            state = preprocess_observation(obs)
            total_reward = 0
            current_steps = 0
        
        # Actor evaluates what to do
        q_values, summary = sess.run([actor_q_values, actor_summary], feed_dict={X_state: [state]})
        file_writer.add_summary(summary, step)
        #print(q_values)
        action_index = epsilon_greedy(q_values, step)
        action = action_index_to_action(action_index)
        #print(action)
        
        #render_env(env)
        # Actor takes action
        obs, reward, done, info = env.step(action)
        if done:
            print("Agent died: current_steps", current_steps, "total_reward", total_reward)
            render_env(env)
        elif (150 < current_steps and current_steps % 10 == 0) or step > 90000:
            render_env(env)
        total_reward += reward
        current_steps += 1
        next_state = preprocess_observation(obs)
        
        # Memorise action.
        replay_memory.append((state, action_index, reward, next_state, 1.0 - done))
        state = next_state
        
        if iteration < training_start or iteration % training_interval != 0:
            continue
        
        # Train the critic.
        X_state_val, X_action_index_val, rewards, X_next_state_val, continues = (
            sample_memory(batch_size))
        next_q_values = actor_q_values.eval(feed_dict={X_state: X_next_state_val})
        max_next_q_values = np.max(next_q_values, axis=1, keepdims=True)
        # Calculate the Actor's reward + future discounted estimated Q-value.
        #print("rewards", rewards.shape, "continues", continues.shape, "max_next_q_values", max_next_q_values.shape)
        y_val = rewards + continues * discount_rate * max_next_q_values
        #print(X_state, X_action_index, y)
        #print("X_state_val", X_state_val.shape, "X_action_index", X_action_index_val.shape, "y_val", y_val.shape)
        training_op.run(feed_dict={X_state: X_state_val,
                                  X_action_index: X_action_index_val,
                                  y: y_val})
        
        if step % copy_steps == 0:
            copy_critic_to_actor.run()
        
        if step % save_steps == 0:
            saver.save(sess, checkpoint_path)

Agent died: current_steps 85 total_reward -6.476626205024618
Agent died: current_steps 42 total_reward -10.239901858949413
Agent died: current_steps 60 total_reward -2.0543943202539077
Agent died: current_steps 68 total_reward -17.565393702150008
Agent died: current_steps 1599 total_reward -122.18657861847262
Agent died: current_steps 1599 total_reward -120.91493402304148
Agent died: current_steps 51 total_reward -18.877167163921523
Agent died: current_steps 1599 total_reward -116.83633279889213
Agent died: current_steps 1599 total_reward -135.10966720016427
Agent died: current_steps 34 total_reward -8.303397609007856
Agent died: current_steps 63 total_reward -16.783349152429647
Agent died: current_steps 1471 total_reward -137.187065940012
Agent died: current_steps 82 total_reward -22.65734494736729
Agent died: current_steps 1599 total_reward -109.58809775914712
Agent died: current_steps 70 total_reward -21.175027260713893
Agent died: current_steps 57 total_reward -4.1809118406636
Agen

In [ ]:
print("Done")

For:
neurons_per_layer = [40, 10]
learning_rate = 0.001

The RL agent learnt how to get an average reward of 3 +- 2. I assume it is learning how to fall over. This is good progress. Running the agent with these hyper paremeters again, did not produce these same results unfortunately.

Rendering the environment has shown that a reward of 3-5 is probably just the agent falling as far forward as possible. Dumb robot.

After rendering the environment every 50 frames, I can see that it either falls over quickly or it splits it legs and very slowly inches forward and then gets killed after 1599 steps.

I couldn't get the agent to learn how to walk and when I checked my code, I realised that I was using the Q-values from the agent as the torque values for it's action, which basically made it stateless (I think).

Watching the agent, I could see that it was using too much force and flipping itself when it wanted to take a step. Let's add some granularity to the torque to see how it goes.